config/memory_persister.go

```diff
@@ -4,56 +4,67 @@
 package config
 
 import (
-	"bytes"
-	"io"
-	"io/ioutil"
+	pb "github.com/square/quotaservice/protos/config"
+	"sync"
+	"github.com/juju/errors"
 )
 
 type MemoryConfigPersister struct {
 	config  string
-	configs map[string][]byte
+	configs map[string]*pb.ServiceConfig
 	*Notifier
+	*sync.RWMutex
 }
 
 func NewMemoryConfigPersister() ConfigPersister {
 	p := &MemoryConfigPersister{
-		configs:  make(map[string][]byte),
-		Notifier: NewNotifier()}
+		configs:  make(map[string]*pb.ServiceConfig),
+		Notifier: NewNotifier(),
+		RWMutex: &sync.RWMutex{}}
 
 	p.Notify()
 	return p
 }
 
-// PersistAndNotify persists a marshalled configuration passed in.
-func (m *MemoryConfigPersister) PersistAndNotify(marshalledConfig io.Reader) error {
-	b, err := ioutil.ReadAll(marshalledConfig)
-	if err != nil {
-		return err
+// PersistAndNotify persists a configuration passed in.
+func (m *MemoryConfigPersister) PersistAndNotify(oldHash string, cfg *pb.ServiceConfig) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// Optimistic lock failure
+	if m.config != oldHash && oldHash != initialHash {
```

Comment:
You need to hash `m.config` here, right?

Reply:
Also, why are we alway comparing to the `initialHash`? I couldn't see that getting updated anywhere.

Reply:
Ignore my first comment, I see that `m.config` is actually the hash. Can you update that to be `configHash`?

Reply:
Removed this for now - I'll add the optimistic version check for all persisters in a separate PR.